### DISCUSSION TITLE
Shovel API documentation needs to be updated?

### DISCUSSION THREAD

**Discussion by nesonger** [rabbitmq/rabbitmq-website#2018] [General]
We are using an API call to get the status of shovels. It appears that the website documentation needs to be updated to match the github page. 
To view the status of a specific shovel, the page https://www.rabbitmq.com/docs/shovel-dynamic shows GET /api/shovels/vhost/{vhost}/{name} {vhost} and this does not work correctly. I am getting back every shovel if the named shovel exists in 3.12, and every shovel in 3.13 even if the named shovel does not exist.

The github page says to see the status of a specific shovel, use GET /api/parameters/shovel/{vhost}/{name}. This does work correctly in 3.12 and 3.13. 

This page has references to the shovel api that don't work:
https://www.rabbitmq.com/docs/shovel-dynamic ->https://github.com/rabbitmq/rabbitmq-website/blob/main/versioned_docs/version-3.13/shovel-dynamic.md

The api examples on this page work correctly but does not match what is on https://www.rabbitmq.com/docs/shovel-dynamic:
https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_shovel_management

**Comment by michaelklishin** [comment]
I'm looking into it.

**Comment by michaelklishin** [comment]
I'm not sure what

```
GET /api/shovels/vhost/{vhost}/{name} {vhost}
```

is meant to say. The [endpoints](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl#L22-L25) are:

 * `GET /api/shovels`
 * `GET /api/shovels/{vhost}`
 * `GET /api/shovels/{vhost}/{name}`
 * `DELETE /api/shovels/{vhost}/{name}`
 * `DELETE /api/shovels/{vhost}/{name}/restart`

And the docs list those endpoints. However,

```
GET /api/shovel/vhost/{vhost}/{shovel}
```

incorrectly returns a list of all shovels instead of just one. The `DELETE` endpoints are not
affected because they do not produce a response.

**Reply by michaelklishin** [reply]
Unless you have a lot of shovels and listing them produces large responses, a good enough solution would be to find the resulting shovel by name in the `GET /api/shovels/{vhost}` response.

Querying parameters will also work but you are not using the Shovel API, just a direct runtime parameters one where `shovel` is a parameter component (type) name.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-server/issues/12040

**Comment by nesonger** [comment]
Then there may be an issue with the shovel management endpoint with 3.13 installed on Windows Server.GET /api/shovelsThis returns all shovelsGET /api/shovels/{vhost}This returns all shovels in the named vhostGET /api/shovels/{vhost}/{name}This returns a 400 error bad syntax GET /api/shovels/vhost/{vhost}/{name}This returns all shovels even when the shovel specified in {name} does not exist. In 3.12 at least it would return a 404 if the shovel does not exist. Now in 3.13 it returns a 200 with all of the shovels.The only way we can get our process to work correctly now is to use the /api/parameters/shovel endpoint. This is what is listed on the two GitHub links below.Now which of these pages are correct for getting the shovel status? Shouldn’t the information be the same on these pages regarding the HTTP API syntax?Configuring Dynamic Shovelsrabbitmq.comrabbitmq/rabbitmq-shovel-management: RabbitMQ Shovel Managementgithub.comrabbitmq-server/deps/rabbitmq_shovel_management at main · rabbitmq/rabbitmq-servergithub.comSent from my iPhoneOn Aug 16, 2024, at 7:36 PM, Michael Klishin ***@***.***> wrote:﻿
I'm not sure what
GET /api/shovels/vhost/{vhost}/{name} {vhost}

is meant to say. The endpoints are:

GET /api/shovels
GET /api/shovels/{vhost}
GET /api/shovels/{vhost}/{name}
DELETE /api/shovels/{vhost}/{name}
DELETE /api/shovels/{vhost}/{name}/restart

And the docs list those endpoints. However,
GET /api/shovel/vhost/{vhost}/{shovel}

incorrectly returns a list of all shovels instead of just one. The DELETE endpoints are not
affected because they do not produce a response.

**Reply by michaelklishin** [reply]
@nesonger I have linked to https://github.com/rabbitmq/rabbitmq-server/issues/12040 above.

**Reply by michaelklishin** [reply]
There is nothing wrong with querying runtime parameters, or you can filter the list of shovels which is a very straightforward JSON object.

The `rabbitmq_shovel_management` plugin never really provided a way to query just one shovel. The single shovel `GET` endpoint was added to the docs by mistake.

**Comment by nesonger** [comment]
Thank you for the clarification and details. I will pass along the information to my team. Sent from my iPhoneOn Aug 16, 2024, at 9:17 PM, Michael Klishin ***@***.***> wrote:﻿
There is nothing wrong with querying runtime parameters, or you can filter the list of shovels which is a very straightforward JSON object.
The rabbitmq_shovel_management plugin never really provided a way to query just one shovel. The single shovel GET endpoint was added to the docs by mistake.
